### ISSUE TITLE
Layout Module open discussion

### ISSUE THREAD

**Issue by baruchvlz** [orchestratora/orchestrator#8]
In order to provide a user configurable layout of `RenderItem[]` we should provide a `LayoutModule` that exports wrappers. One wrapper should be a linear flex wrapper `FlatLayoutComponent` and the other should a grid wrapper `GridLayoutComponent`.

I suggest that in order to complete this as fast a possible, we integrate these wrapper using `@angular/flex-layout` and exposing basic configurations  for each wrapper.

_

**FlatLayoutComponent**

The flat layout should wrap `n` amount of `OrchestratorConfigItem` instances and emit an event when all the instances have been rendered. It should also have a default configuration for certain properties required for `@angular/flex-layout`.

Consider this as a default config (note that these configurations can be overwritten with `Input`s)

```typescript
import { Injectable } from '@angular/core';

@Injectable({ providedIn: 'root' })
export class OrchestratorFlatLayoutConfig {
  direction = 'row';
  wrap = 'wrap';
  align = 'space-between center';
  columns = 3;
}
```

Apart from the configuration `Input`s, the component should also take in items to be rendered `items: ReadonlyArray<OrchestratorConfigItem<any>>;`, should have an event emitter for when the items are rendered and an even emitter for everytime a new item is added into the view.

_

**GridLayoutComponent**

_TBD_

**Comment by gund** [comment]
Great stuff.

A few inputs from me:
- For better discoverability I suggest the we name components with the type being in first place: `FlatLayoutComponent` => `LayoutFlatComponent` etc.
This way users will get more relevant autocomplete results by just typing `Layout` and see what options do they have.
- Bringing in dependency on angular's flex layout might be an overhead in our case. I suggest that we provide basic layouts with pure simple CSS styles and if needed we can also implement same layouts based on angular flex in separate package.

**Comment by baruchvlz** [comment]
Yea, I like `LayoutFlatComponent` better, it provides the package name first then the type of component.

As for the flex layout dependency, I figured we could provide all the responsiveness that comes along with the library, but I'm not opposed to moving this to a separate package to be done later. Maybe after we integrate drag and drop customisation. Basic flex CSS with minimum configuration should be enough for now.

**Comment by baruchvlz** [comment]
Also, I think the basic flex layout should be done through directives in order to keep the CSS inside the project close to none. We could also share these directives with other packages / layout components and keep everything tidy and easy to control.

**Comment by gund** [comment]
But you cannot keep CSS styles in directives since they do not have templates. So have to use components for that.

**Comment by baruchvlz** [comment]
What if we don't use CSS styles at all?

```typescript

@Directive({
  selector: '[orcFlex]',
  exportAs: 'orcFlex',
})
export class FlexDirective {
  @HostBinding('style.display') display = 'flex';
}
```

**Comment by gund** [comment]
Yeah, that's like CSS in directive. It will work. I like the idea of moving stuff into directives.

**Comment by gund** [comment]
Yes, but what I mean is that there is no way to apply directives on dynamic components, so in this case you would not be able to set `FlexDirective` on the layout that is dynamically rendered because you do not control the template.

For that you have to create a wrapper component for a layout that will attach a directive to it, and then you dynamically render that wrapper component but not the layout directly. So you sort of creating a "host" component for your dynamic component.

Also regarding dynamic directives - I've done some work on it in [here](https://github.com/gund/ng-dynamic-component/pull/173) but there are limitations on what is supported from those dynamic directives.

**Comment by gund** [comment]
@RyuuGan as we already discussed we do not want to implement everything that angular flex does, that is why we also do not want to depend on it, because it is just too much for a simple flex with a bunch of CSS properties.

But you made me think about should we really do it via directives, it has only one benefit - being reusable (but do we really need it), however with directives CSS styles will be applied as `style` attribute which is far from ideal, and it will be harder for users to override those styles from CSS.

So maybe we should simply write a few CSS props right in component, and only those needed configuration provide as directives?

**Comment by RyuuGan** [comment]
@gund That is what I wanted to propose too. I don't like inline styling too because it is hard to override. And at the beginning we need only a flex component that implements rows. We don't need columns right now because it is a default behavior of block components (of course it is a little bit different).

So I guess the simple `LayoutRowComponent` for the beginning is quite enough, maybe event without row wrapping.

**Comment by gund** [comment]
Yes, but it's better to have more flexible component which supports `row` and `column` directions, since it's just one CSS property to change it is very easy to implement.

So we can have `LayoutFlatComponent` which will have most of it CSS in component but some as directive to be configurable, like `flexDirection` with a default value to be `row`.

For this purpose it's okay to set CSS via styles because user will have control over it from the component inputs so there will be no need to override CSS directly.

**Comment by baruchvlz** [comment]
If we don't want to use inline styling, we could have pre-set classes that get added to the element depending on props passed. This way it'll be easier to override with specificity. And those styles that shouldn't be overwritten can be added as inline styles.

I think the 4 options in the Config I posted should be enough to manage the basic flex layout. Any other customization can be done by the user or they could use the more editable `LayoutGridComponent`.

**Comment by baruchvlz** [comment]
So, the tests I have for this so far are the following. If you think I should add any let me know, they aren't many.

```typescript
  it('should render n amount of items', () => {});

  it('should apply the correct default styles to the wrapper', () => {});

  it('should emit an event after all items are rendered', () => {});
```

By default the wrapper will have these styles

```css
section.layout-flat {
  display: flex;
  width: 100%;
}
```

And the component will have these input properties

```typescript
  @Input() items: ReadonlyArray<OrchestratorConfigItem<any>>;
  @Input() direction: LayoutFlatDirectionOptions;
  @Input() wrap: LayoutFlatWrapOptions;
  @Input() justify: LayoutFlatJustifyOptions;
  @Input() alignItems: LayoutFlatAlignItemsOptions;
  @Input() alignContent: LayoutFlatAlignContentOptions;
  @Input() columns: number;
```

**Comment by gund** [comment]
@baruchvlz all looks reasonable.

What I think is why wouldn't we make the host component being `flex` itself?
This way we can get rid of one extra inner element (`section`).

**Comment by gund** [comment]
This RFC was migrated to https://github.com/orchestratora/rfcs/blob/master/implemented/0001-layout-module.md
